### PR TITLE
active_support/indifferent_access: fix not raising when default_proc does

### PR DESCRIPTION
**Pull request by sirupsen**
Ran into this with our Rails application: `HashWithDifferentAccess`'s behaviour is not consistent with Ruby when `default_proc` is set and cloned and when it raises.

``` ruby
h = HashWithIndifferentAccess.new
h.default_proc = proc { |h, k| raise "OMG" }
h.dup
# => RuntimeError: OMG
```

``` ruby
h = {}
h.default_proc = proc { |h, k| raise "OMG” }
h.dup
# => {}
```

This PR fixes `#dup` and `#to_hash` to set defaults safely when `#default_proc` raises, by avoiding to call into it.

\cc @byroot @arthurnn @rafaelfranca

### CODE REVIEW

File: activesupport/lib/active_support/hash_with_indifferent_access.rb [hunk 1]
```diff
@@ -275,6 +277,14 @@ def convert_value(value, options = {})
           value
         end
       end
+
+      def set_defaults(target, source = self)
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/hash_with_indifferent_access.rb]: What is the reason for the source argument?

Comment (Sirupsen) [inline activesupport/lib/active_support/hash_with_indifferent_access.rb]: I originally used this for the `new_from_hash_copying_default` as well, but realized later I didn't need it. I'll remove it.

**Comment by rafaelfranca** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by Sirupsen** [comment]
Thanks @rafaelfranca!

**Comment by rafaelfranca** [comment]
Backported in 9ce5a79

**Pull request merged**